src/server/pfs/fuse/server.go

```diff
@@ -1210,13 +1225,166 @@ func sanitizeInputAndGetAlias(datumInput *pps.Input, c *client.APIClient) (map[s
 		}
-
 		return nil
 	}); err != nil {
+		return err
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.DatumInput = datumInput
+		mm.DatumInputsToMounts = datumInputsToMounts
+	}()
+	return nil
+}
+
+func (mm *MountManager) GetNextXDatums(numDatums int, paginationMarker string) ([]*pps.DatumInfo, error) {
+	req := &pps.ListDatumRequest{
+		Input:            mm.DatumInput,
+		Number:           int64(numDatums),
+		PaginationMarker: paginationMarker,
+	}
+
+	ctx, cf := context.WithCancel(mm.Client.Ctx())
+	defer cf()
+	client, err := mm.Client.PpsAPIClient.ListDatum(ctx, req)
+	if err != nil {
 		return nil, err
 	}
+	var dis []*pps.DatumInfo
+	for {
+		di, err := client.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return dis, nil
+			}
+			return nil, err
+		}
+		dis = append(dis, di)
+	}
+}
 
-	return datumInputsToMounts, nil
+func (mm *MountManager) GetDatums() (retErr error) {
+	defer func() {
+		retErr = grpcutil.ScrubGRPC(retErr)
+	}()
+	// TODO: Try creating datums using GlobFile for now because it's faster than ListDatum.
+	// Once ListDatum pagination is efficient, we can remove this. #ListDatumPagination
+	if err := mm.CreateDatums(); err == nil {
+		return nil
+	} else if !errors.Is(err, ErrUnsupportedInputType) {
+		return err
+	}
+	// If input spec has Join or Group, fall back to ListDatum
+	datums, err := mm.GetNextXDatums(NumDatumsPerPage, "")
+	if err != nil {
+		return err
+	}
+	if len(datums) == 0 {
+		return errors.New("no datums to mount")
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.Datums = datums
+		mm.PaginationMarker = datums[len(datums)-1].Datum.Id
+		if len(datums) < NumDatumsPerPage || NumDatumsPerPage == 0 {
+			mm.AllDatumsReceived = true
+		}
+	}()
+	return nil
+}
+
+func visitInput(input *pps.Input, level int, f func(*pps.Input, int) error) error {
+	var source []*pps.Input
+	switch {
+	case input == nil:
+		return errors.Errorf("spouts not supported") // Spouts may have nil input
+	case input.Cross != nil:
+		source = input.Cross
+	case input.Join != nil:
+		source = input.Join
+	case input.Group != nil:
+		source = input.Group
+	case input.Union != nil:
+		source = input.Union
+	}
+	for _, input := range source {
+		if err := visitInput(input, level+1, f); err != nil {
+			return err
+		}
+	}
+	return f(input, level)
+}
+
+func crossDatums(datums [][]*pps.DatumInfo) []*pps.DatumInfo {
+	if len(datums) == 0 {
+		return nil
+	}
+	ret := datums[0]
+	for _, slice := range datums[1:] {
+		var temp []*pps.DatumInfo
+		for _, d1 := range ret {
+			for _, d2 := range slice {
+				temp = append(temp, &pps.DatumInfo{
+					Data: append(d1.Data, d2.Data...),
+				})
+			}
+		}
+		ret = temp
+	}
+	return ret
+}
+
+func (mm *MountManager) CreateDatums() error {
+	datumsAtLevel := make(map[int][][]*pps.DatumInfo)
+	if err := visitInput(mm.DatumInput, 0, func(input *pps.Input, level int) error {
+		if input.Cron != nil || input.Join != nil || input.Group != nil {
```

Comment:
it seems like given the right input spec we could be doing a lot of wasted work and making a lot of globfile calls before we break out after encountering one of these input types

Reply:
Good point. We traverse the input spec once earlier in a less expensive fashion, so I'll use that to decide whether we use GlobFile or ListDatum.